util/occamygen.py

```diff
@@ -437,13 +469,92 @@ def main():
     soc_regbus_periph_xbar.add_output_entry("hbm_phy_cfg", am_hbm_phy_cfg)
     soc_regbus_periph_xbar.add_output_entry("hbm_seq", am_hbm_seq)
 
-    #################
-    # SoC Wide Xbar #
-    #################
+    ##################
+    # SoC Wide Xbars #
+    ##################
+
+    # Quadrant pre xbars
+    # Each connects one quadrant and one HBI master to the HBM and quadrant xbars
+    quadrant_pre_xbars = list()
+    for i in range(nr_s1_quadrants):
+        quadrant_pre_xbar = solder.AxiXbar(
+            48,
+            512,
+            occamy.cfg["pre_xbar_slv_id_width_no_rocache"] + (
+                1 if occamy.cfg["s1_quadrant"].get("ro_cache_cfg") else 0),
+            name="quadrant_pre_xbar_{}".format(i),
+            clk="clk_i",
+            rst="rst_ni",
+            max_slv_trans=occamy.cfg["quadrant_pre_xbar"]["max_slv_trans"],
+            max_mst_trans=occamy.cfg["quadrant_pre_xbar"]["max_mst_trans"],
+            fall_through=occamy.cfg["quadrant_pre_xbar"]["fall_through"],
+            no_loopback=True,
+            atop_support=False,
+            context="soc",
+            node=am_quadrant_pre_xbar[i])
+
+        # Default port:
+        quadrant_pre_xbar.add_output_entry("quadrant_inter_xbar", am_quadrant_inter_xbar)
+        quadrant_pre_xbar.add_output_entry("hbm_xbar", am_hbm_xbar)
+        quadrant_pre_xbar.add_input("quadrant")
+
+        quadrant_pre_xbars.append(quadrant_pre_xbar)
+
+    # Quadrant inter xbar
+    # Connects all quadrant pre xbars to all quadrants, with additional wide xbar M/S pair
+    quadrant_inter_xbar = solder.AxiXbar(
+        48,
+        512,
+        quadrant_pre_xbars[0].iw_out(),
+        name="quadrant_inter_xbar",
+        clk="clk_i",
+        rst="rst_ni",
+        max_slv_trans=occamy.cfg["quadrant_inter_xbar"]["max_slv_trans"],
+        max_mst_trans=occamy.cfg["quadrant_inter_xbar"]["max_mst_trans"],
+        fall_through=occamy.cfg["quadrant_inter_xbar"]["fall_through"],
+        no_loopback=True,
+        atop_support=False,
+        context="soc",
+        node=am_quadrant_inter_xbar)
+
+    # Default port: soc wide xbar
+    quadrant_inter_xbar.add_output_entry("wide_xbar", am_soc_wide_xbar)
+    quadrant_inter_xbar.add_input("wide_xbar")
+    for i in range(nr_s1_quadrants):
+        # Default route passes HBI through quadrant 0
+        # --> mask this route, forcing it through default wide xbar
+        quadrant_inter_xbar.add_output_entry("quadrant_{}".format(i),
+                                             am_wide_xbar_quadrant_s1[i],
+                                             range_mask=(0, 0x1000000000))
+        quadrant_inter_xbar.add_input("quadrant_{}".format(i))
```

Comment:
A short clarification question: so by default it would have routed through Quadrant 0 to the NarrowXbar to the Narrrow Die2Die link?

Reply:
Aah, good catch. This is a leftover from when every quadrant had its own HBI port and the routing algorithm of solder would insist on routing all HBI traffic through quadrant 0 in the inter xbar (fewer hops). Since this is no longer possible, this fix is redundant now.